src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProjectFactory.cs

```diff
@@ -78,57 +77,76 @@ public ProjectSystemProjectFactory(Workspace workspace, IFileChangeWatcher fileC
+                : VersionStamp.Create();
+
+            var projectInfo = ProjectInfo.Create(
+                new ProjectInfo.ProjectAttributes(
+                    projectId,
+                    versionStamp,
+                    name: projectSystemName,
+                    assemblyName,
+                    language,
+                    compilationOutputFilePaths: default, // will be updated when command line is set
+                    SourceHashAlgorithms.Default, // will be updated when command line is set
+                    filePath: creationInfo.FilePath,
+                    telemetryId: creationInfo.TelemetryId),
                 compilationOptions: creationInfo.CompilationOptions,
-                filePath: creationInfo.FilePath,
                 parseOptions: creationInfo.ParseOptions);
 
-            var versionStamp = creationInfo.FilePath != null ? VersionStamp.Create(File.GetLastWriteTimeUtc(creationInfo.FilePath))
-                                                             : VersionStamp.Create();
-
-            await ApplyChangeToWorkspaceAsync(w =>
+            await ApplyChangeToWorkspaceAsync(async w =>
             {
-                var projectInfo = ProjectInfo.Create(
-                    new ProjectInfo.ProjectAttributes(
-                        id,
-                        versionStamp,
-                        name: projectSystemName,
-                        assemblyName: assemblyName,
-                        language: language,
-                        checksumAlgorithm: SourceHashAlgorithms.Default, // will be updated when command line is set
-                        compilationOutputFilePaths: default, // will be updated when command line is set
-                        filePath: creationInfo.FilePath,
-                        telemetryId: creationInfo.TelemetryId),
-                    compilationOptions: creationInfo.CompilationOptions,
-                    parseOptions: creationInfo.ParseOptions);
-
-                // If we don't have any projects and this is our first project being added, then we'll create a new SolutionId
-                // and count this as the solution being added so that event is raised.
-                if (w.CurrentSolution.ProjectIds.Count == 0)
-                {
-                    w.OnSolutionAdded(
-                        SolutionInfo.Create(
-                            SolutionId.CreateNewId(SolutionPath),
-                            VersionStamp.Create(),
-                            SolutionPath,
-                            projects: new[] { projectInfo },
-                            analyzerReferences: w.CurrentSolution.AnalyzerReferences)
-                        .WithTelemetryId(SolutionTelemetryId));
-                }
-                else
-                {
-                    w.OnProjectAdded(projectInfo);
-                }
+                await w.SetCurrentSolutionAsync(
```

Comment:
this now uses the new helper, spinning until it 'wins' and actually makes the change to the workspace.